pkg/controller/master/upgrade/upgrade_controller.go

```diff
@@ -723,3 +750,199 @@ func (h *upgradeHandler) setReplicaReplenishmentValue(value int) error {
 	}
 	return nil
 }
+
+// storeVMState stores VM state only in single-node environments.
+// This method stores VM only when Ready=true, meaning it only saves VMs info
+// in running state to the secret.
+func (h *upgradeHandler) storeVMState(upgrade *harvesterv1.Upgrade) error {
+	isSingleNodeRestoreVM, err := h.isSingleNodeRestoreVM(upgrade)
+	if err != nil {
+		return err
+	}
+	if !isSingleNodeRestoreVM {
+		logrus.WithFields(logrus.Fields{
+			"namespace": upgrade.Namespace,
+			"name":      upgrade.Name,
+		}).Info("Skip store VM state")
+		return nil
+	}
+
+	vms, err := h.vmCache.List(corev1.NamespaceAll, labels.Everything())
+	if err != nil {
+		logrus.WithFields(logrus.Fields{
+			"namespace": upgrade.Namespace,
+			"name":      upgrade.Name,
+		}).WithError(err).Error("Failed to list all VMs")
+		return err
+	}
+
+	runningVMs := make([]*PreUpgradeRunningVM, 0)
+	for _, vm := range vms {
+		if vmReady.IsTrue(vm) {
+			vm := &PreUpgradeRunningVM{
+				Namespace: vm.Namespace,
+				Name:      vm.Name,
+			}
+			runningVMs = append(runningVMs, vm)
+		}
+	}
+	runningVMsJSON, err := json.Marshal(runningVMs)
+	if err != nil {
+		logrus.WithFields(logrus.Fields{
+			"namespace": upgrade.Namespace,
+			"name":      upgrade.Name,
+		}).WithError(err).Error("Failed to marshal PreUpgradeRunningVM list to json")
+		return err
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: harvesterSystemNamespace,
+			Name:      preUpgradeVMsSecretName,
+		},
+		Data: map[string][]byte{
+			"vms": runningVMsJSON,
+		},
+	}
```

Comment:
Do we want to add an owner reference for the Secret? If the user removes the Upgrade object in the middle of an upgrade, the Secret without the Upgrade object as its owner will remain on the system.

Reply:
Make sense to me, will add the owner reference. Thanks !

Reply:
Already addressed all comments in the latest commit.